src/common/storage/src/stage.rs

```diff
@@ -29,3 +50,185 @@ pub fn init_stage_operator(stage_info: &UserStageInfo) -> Result<Operator> {
+                    pattern, e
+                ))),
+            },
+            None => Ok(None),
+        }
+    }
+
+    pub async fn list(&self, operator: &Operator, first_only: bool) -> Result<Vec<FileWithMeta>> {
+        if let Some(files) = &self.files {
+            let mut res = Vec::new();
+            for file in files {
+                let full_path = Path::new(&self.path)
+                    .join(file)
+                    .to_string_lossy()
+                    .to_string();
+                let meta = operator.object(&full_path).metadata().await?;
+                if meta.mode().is_file() {
+                    res.push(FileWithMeta::new(&full_path, meta))
+                } else {
+                    return Err(ErrorCode::BadArguments(format!(
+                        "{full_path} is not a file"
+                    )));
+                }
+                if first_only {
+                    break;
+                }
+            }
+            Ok(res)
+        } else {
+            let pattern = self.get_pattern()?;
+            list_files_with_pattern(operator, &self.path, pattern, first_only).await
+        }
+    }
+
+    pub async fn first_file(&self, operator: &Operator) -> Result<FileWithMeta> {
+        let mut files = self.list(operator, true).await?;
+        match files.pop() {
+            None => Err(ErrorCode::BadArguments("no file found")),
+            Some(f) => Ok(f),
+        }
+    }
+
+    pub fn blocking_first_file(&self, operator: &Operator) -> Result<FileWithMeta> {
+        let mut files = self.blocking_list(operator, true)?;
+        match files.pop() {
+            None => Err(ErrorCode::BadArguments("no file found")),
+            Some(f) => Ok(f),
+        }
+    }
+
+    pub fn blocking_list(
+        &self,
+        operator: &Operator,
+        first_only: bool,
+    ) -> Result<Vec<FileWithMeta>> {
+        if let Some(files) = &self.files {
+            let mut res = Vec::new();
+            for file in files {
+                let full_path = Path::new(&self.path)
+                    .join(file)
+                    .to_string_lossy()
+                    .to_string();
+                let meta = operator.object(&full_path).blocking_metadata()?;
+                if meta.mode().is_file() {
+                    res.push(FileWithMeta::new(&full_path, meta))
+                } else {
+                    return Err(ErrorCode::BadArguments(format!(
+                        "{full_path} is not a file"
+                    )));
+                }
+                if first_only {
+                    break;
+                }
+            }
+            Ok(res)
+        } else {
+            let pattern = self.get_pattern()?;
+            blocking_list_files_with_pattern(operator, &self.path, pattern, first_only)
+        }
+    }
+}
+
+async fn list_files_with_pattern(
+    operator: &Operator,
+    path: &str,
+    pattern: Option<Regex>,
```

Comment:
Should we use `glob::Pattern` or  `Regex`?

`glob::Pattern` will follow Unix shell style, and it is different from regex.

For example: `*.parquet`,

- glob: will match all files with a suffix '.parquet'.
- regex: invalid.

Reply:
FYI:  we used to use  Regex in copy
https://docs.snowflake.com/en/sql-reference/sql/copy-into-table.html#optional-parameters